Repository Configuration

tajmone opened the pull request:
- Add EditorConfig settings to enforce code consistency.
- Enforce code consistency across repository files.
- Tweak `.gitattributes` for better cross-platform interoperability  and for Windows end-users that don't have auto-CRLF enabled;  also add some GH Linguist directives for unknown syntaxes.
- Tweak `.gitignore` for better cross-platform interoperability,  adding exclusion patterns for OS specific auto-generated system files.
- Add BOM to all test files, to enforce consistency (some had a BOM  while others didn't) and for better interoperability with PB IDE.

peterthomashorn commented on .gitignore:
```diff
@@ -1 +1,84 @@
+# Recycle Bin used on file shares
+$RECYCLE.BIN/
+
+# Windows Installer files
```
At one hand I feel unsure about the need of some exclusions, especially these. On the other hand I know that they do no harm, too. Such binaries probably are better kept in the global ignore file, I would guess.

tajmone replied:
When it comes to repository configurations I prefer the "set it and forge it" approach, i.e. covering all general occurrences and common repository patterns, so that if new contents are submitted the settings are all in place.

E.g. I've added all the known ST file extensions to the config, even if we don't use all of them right now, so if we add new files we don't have to tweak the configuration every time.

> Such binaries probably are better kept in the global ignore file, I would guess.

They are added to prevent unwanted files in third party pull requests, especially in case of Git newbies. We can't make any assumption on the Git and machine settings of contributors.

These patterns were taken from the gitignore project which advises best practices for cross-platform projects, covering all the common ignore patterns for the major OSs, including old versions of Windows. The idea is to prevent setting files from ending up in commits, especially in third party contributions. 

That's why I think it's best to have them in the repository configuration, to be on the safe side — they are harmless, but they can prevent garbage ending up in contributions.

Because this is a repository related to PureBasic, I'm adding extra comments and taking extra precautions, because for the PB community using Git is something relatively new, prompted by the IDE source having been released on GitHub.

In the official PB IDE repository, we've seen a lot of pull requests with whitespace noise, system files, etc., and we had to take extra measures to enforce CI validation to keep contributions clean.

peterthomashorn commented on .gitattributes:
```diff
@@ -1,6 +1,56 @@
 *.sublime-project export-ignore
+.editorconfig export-ignore
+/.git* export-ignore
+/.github export-ignore
```
Maybe because it is 22:18 here or I am new to `.gitattributes` but isn't this the usual shell globbing pattern? Isn't line 53 already covered by line 52?

tajmone replied:
Indeed, it's redundant. Fixed now.

peterthomashorn commented on Tests/syntax_test_pointer.pb:
```diff
@@ -113,4 +113,4 @@ Define *StringAddress = @"This is a string in memory"
 ; Folding = -
-; EnableXP
\ No newline at end of file
+; EnableXP
```
I updated this test file to not contain the PureBasic IDE preferences anymore.

tajmone replied:
In the PB IDE repository I've added a custom script that validates commits and PRs for code styles consistency, and checks PB files to ensure they don't contain settings saved in the source:

[`validate.sh`][validate.sh]


That's a Bash script, which I don't like much but at the time was my best option. I could recreate the PB settings validator as a Ruby script and add it to this repository, we could then use it for CI validation and as a Git pre-commit hook to, so checks are done locally too. 

In the PB repository I can't use Ruby because it would add an extra dependency to the project, but here we could use Ruby and Rake for all automation task.

<!----------------------------- REFERENCE LINKS ------------------------------>

[validate.sh]: https://github.com/fantaisie-software/purebasic/blob/devel/validate.sh

<!-- EOF -->

peterthomashorn reviewed (approved):
I see no blocker. Consider my comments as suggestions.

peterthomashorn commented:
Oops. Now I messed up something with my removal of the PureBasic IDE preferences at the end of the test file. This is the first time I use GitHub for reviewing a pull request to something I maintain. A rebase and a simple conflict resolution of that file should be sufficient.

peterthomashorn commented:
Just now I had the idea that the EditorConfig utilization might be worth mentioning in the README. As far as I know it also requires the third-party package to be installed. That is not something users might be aware of. In fact I did not know about EditorConfig until a few weeks ago. I would consider it a feature because it takes care about some things.

tajmone commented:
> Oops. Now I messed up something with my removal of the PureBasic IDE preferences at the end of the test file.

It's fine, I've rebased the commit and solved all conflicts using `--ours`.

> Just now I had the idea that the EditorConfig utilization might be worth mentioning in the README. As far as I know it also requires the third-party package to be installed.

On many modern editors it's supported out of the box, whereas others require a third party plug-in (e.g. Sublime Text).

> That is not something users might be aware of. In fact I did not know about EditorConfig until a few weeks ago.

It might be worth mentioning in the CONTRIBUTING document that we use EditorConfig for code style consistency and provide a link to the official repository, which also lists all the plugins, etc. I've already written this text in many repositories, so I can reuse the paragraph and paste it in `CONTRIBUING.md`.


> I would consider it a feature because it takes care about some things.

In most repositories I use the `.editorconfig` settings to validate commits and PRs for code consistency; we also use it on the PB IDE repository to filter out noise from PRs — you wouldn't believe how many PRs add noise such as switching from space indentations to tabs every changed file, and other aesthetic changes which make it impossible to see read through the commits diffs.

Unfortunately, the validation tool I've been using so far is a bit buggy and has been discontinued for a long time. I've found a replacement tool, but it had a bug that prevented using it on Windows; the bug was now fixed, and I'm just waiting that the Node package is updated with the new version then I'll switch all my repositories to use the new validator (it's a matter of weeks and it should be ready).

Locally, I use a tweaked version of the validation script as a pre-commit Git hook to check all modified files for code consistency, so I don't push cluttered contents on my own repositories.

tajmone commented:
# Updated `CONTRIBUTING.md`

I've updated `CONTRIBUTING.md`, as you suggested, mentioning the use of EditorConfig to enforce code styles consistency, and providing a link to the EditorConfig package for Sublime Text.

I've also pasted over the instructions that I wrote for the PB IDE repository on how to correctly set the PD IDE to prevent settings from being saved in the tests sources files (see: [original doc on PB repos]).


<!----------------------------- REFERENCE LINKS ------------------------------>

[original doc on PB repos]: https://github.com/fantaisie-software/purebasic/blob/devel/CONTRIBUTING.md#purebasic-ide-settings

<!-- EOF -->
